Apply Humble object pattern on Items

TaylorNAlbarnaz opened the pull request:
## Summary

Apply a soft version of the humble object pattern to allow tests on items, containers and roles. Separating network/unity and item logic. 
This PR does the following : 

- Move some unity/network logic out of container and back on attachedContainer.
- Add some [server] [client] attributes where it's needed.
- Rework item to make it's traits list sync, as well as it's container (using syncObjects).
- added some commands for testing traits in play mode
- added some tests for items and containers
- modify fishnet to allow editor tests on some netcode stuff
- added a constructor-like method in item : Init()

## Changes to Files

Most changes are in item, container and attached container. 
Other changes are minor update to fit the new item class.
Item prefabs were updated to fit with the new item script.

## TODO

- [x]  write 1-2  tests for items
- [x]  write 1-2 tests for containers
- [x] document every major changes
- [x] check in modified code that everything has a correct scope (public, protected, private) 
- [x] write custom serializer for traits
- [x] fix the bug where client can't directly put things in hand when clicking on PDA
- [x] test if it's possible to put [Server] [Client] attributes in humble objects without ruining testing. If yes, do it appropriately.
- [x] add command for describing item in hand.
- [x] Implement Delete in item or remove it.
- [x] make item ref private with public getter in ItemActor 
- [x] make some stuff constant like size, name (made them read only)
- [ ] Write some doc on turning stuff into HO with networking involved.
- [x] fix code style
- [x] remove buggy container syncing using RPC. Replace with syncvar
- [x] mod fishnet for editor tests

TaylorNAlbarnaz committed:
Applies the Humble Object Pattern to Items.cs

TaylorNAlbarnaz committed:
Refactor Item to be inside ItemActor class

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Items/ItemActor.cs:
```diff
@@ -26,62 +26,53 @@ namespace SS3D.Systems.Inventory.Items
+        public string Name => item.Name;
+        public ItemId ItemId => item.ItemId;
+        public Vector2Int Size => item.Size;
+        public List<Trait> traits => item.Traits;
```
replace that with 
```
[SyncObject]
        private readonly SyncList<Trait> traits = new();
```

add the property : 

`public ReadOnlyCollection<Trait> Traits => ((List<Trait>) traits.Collection).AsReadOnly();`

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Items/ItemActor.cs:
```diff
@@ -91,20 +82,24 @@ public Sprite InventorySprite
+            item.Traits = _startingTraits;
+
+            item.Sprite = null;
 
```
Empty the starting traits list, set starting name string to "", free as much memory as possible here.

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Items/ItemActor.cs:
```diff
@@ -91,20 +82,24 @@ public Sprite InventorySprite
+        public new void Awake()
         {
-            _sprite = null;
+            item.Name = _startingName;
```
use a constructor instead of setters, goal is to keep those fields private and not easily modifiable (with some logic at least, not just freely)

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Items/ItemActor.cs:
```diff
@@ -124,17 +119,31 @@ protected override void OnStart()
 
             // Items can't have no size
-            if (_size.x == 0)
+            if (item.Size.x == 0)
```
that should go in the item constructor

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Items/ItemActor.cs:
```diff
@@ -124,17 +119,31 @@ protected override void OnStart()
+        /// Changes the item actor's item id
+        /// </summary>
+        /// <param name="id">AssetDatabase's ItemId</param>
+        public void SetId(ItemId id)
```
why do we need this ? is that new or was it something in item before ?

TaylorNAlbarnaz replied:
It i just set the id manually it throws a null ref exception, because the AssetDatabase will create an ItemActor and try to access Item before the Awake event is called

stilnat replied:
so if that's a one time thing, I'd rather have it event based, so AssetDatabase wait for ItemActor to be setup. Another possible solution, have a bool like hasAwaken that turn true in itemID, so that it doesn't get called again.

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Items/ItemActor.cs:
```diff
@@ -226,46 +235,44 @@ public override void CreateSourceInteractions(IInteractionTarget[] targets, List
+        /// Checks if the item is currently stored in a container
+        /// </summary>
+        /// <returns></returns>
+        public bool InOnContainer()
```
change to IsOnContainer

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Items/ItemActor.cs:
```diff
@@ -226,46 +235,44 @@ public override void CreateSourceInteractions(IInteractionTarget[] targets, List
         public void SetContainer(Container newContainer, bool alreadyAdded, bool alreadyRemoved)
         {
-            if (_container == newContainer)
+            if (Container == newContainer)
```
that logic can probably go in item

TaylorNAlbarnaz replied:
The logic is server-side, meaning the Item class would be calling Server functions and ServerRPCs, kinda breaks the Humble Object Pattern in my opinion

stilnat replied:
What I mean is something like this : 
in itemActor
```
         [Server]
        public void SetContainer(Container newContainer, bool alreadyAdded, bool alreadyRemoved)
        {
            bool containerSet = item.SetContainer(newContainer, alreadyAdded, alreadyRemoved);
            if (containerSet) RpcSetContainer(newContainer);
        }
```

in item
```
        public bool SetContainer(Container newContainer, bool alreadyAdded, bool alreadyRemoved)
        {
            if (Container == newContainer)
            {
                return false;
            }

            if (Container != null)
            {
                Container.RemoveItem(this);
            }

            if (!alreadyAdded && newContainer != null)
            {
                newContainer.AddItem(this);
            }

            item.Container = newContainer;
            return true;
        }
```

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Items/ItemActor.cs:
```diff
@@ -303,16 +310,107 @@ public void GenerateNewIcon()
+                set => _itemId = value;
+            }
+
+            public string Name
```
make that getter only for now ?

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Items/ItemActor.cs:
```diff
@@ -303,16 +310,107 @@ public void GenerateNewIcon()
+                set => _name = value;
+            }
+
+            public float Weight
```
make that getter only for now ?

TaylorNAlbarnaz replied:
I'll change them to a constructor then

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Items/ItemActor.cs:
```diff
@@ -303,16 +310,107 @@ public void GenerateNewIcon()
+                set => _weight = value;
+            }
+
+            public Vector2Int Size
```
make that getter only for now ?

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Items/ItemActor.cs:
```diff
@@ -303,16 +310,107 @@ public void GenerateNewIcon()
+                set => _sprite = value;
+            }
+
+            public List<Trait> Traits
```
replace with something like public ReadOnlyCollection<Trait> Traits => ((List<Trait>) traits.Collection).AsReadOnly();

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Items/ItemActor.cs:
```diff
@@ -303,16 +310,107 @@ public void GenerateNewIcon()
+                set => _traits = value;
+            }
+
+            public Container Container
```
this one we need to be careful. it wasn't synced properly before because it updated using an RPC, so late joinin client would have the wrong container value when joining. We may need to make Container a syncvar or something (not this one but the one in itemActor)

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Items/ItemSystem.cs:
```diff
@@ -39,9 +39,14 @@ private void LoadItemPrefabs()
-                item._itemId = id;
+                if (item == null)
+                {
+                    item = new ItemActor();
```
I thought you can't instantiate behavior with the new operator ?

TaylorNAlbarnaz replied:
Just an small error in the code, it was supposed to be new Item();

stilnat reviewed (changes requested):
a few remarks on the two main files, the rest will come. Let me know if there's some stuff you disagree with, It's all open to discussion, I'm a bit at sea on those things too, still trying to figure the best way to do it.

TaylorNAlbarnaz committed:
Small Refactor

TaylorNAlbarnaz committed:
Change Traits to be a SyncList

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Containers/Container.cs:
```diff
@@ -627,7 +627,7 @@ private void handleItemRemoved(Item item)
         }
 
-        private void handleItemAdded(Item item)
+        private void handleItemAdded(ItemActor item)
```
that should go in attachedContainer, it's handling Unity logic.

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Containers/Container.cs:
```diff
@@ -646,12 +646,12 @@ private void handleItemAdded(Item item)
         }
 
-        private void HandleContainerContentsChanged(Container container, IEnumerable<Item> oldItems, IEnumerable<Item> newItems, ContainerChangeType type)
+        private void HandleContainerContentsChanged(Container container, IEnumerable<ItemActor> oldItems, IEnumerable<ItemActor> newItems, ContainerChangeType type)
```
that should go in attachedContainer, it's handling Unity logic.

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Containers/Container.cs:
```diff
@@ -608,7 +608,7 @@ public int FindItem(Item item)
         }
 
-        private void handleItemRemoved(Item item)
+        private void handleItemRemoved(ItemActor item)
```
that should go in attachedContainer, it's handling Unity logic.

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Containers/Container.cs:
```diff
@@ -67,7 +67,7 @@ public sealed class Container
         /// The items stored in this container
         /// </summary>
-        public IEnumerable<Item> Items => StoredItems.Select(x => x.Item);
+        public IEnumerable<ItemActor> Items => StoredItems.Select(x => x.Item);
```
that should stay Item.

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Containers/Container.cs:
```diff
@@ -133,7 +133,7 @@ public Container(AttachedContainer attachedContainer)
         /// <param name="item">The item to place</param>
         /// <returns>If the item was added</returns>
-        public bool AddItem(Item item)
+        public bool AddItem(ItemActor item)
```
that should stay item

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Containers/Container.cs:
```diff
@@ -171,7 +171,7 @@ public bool AddItem(Item item)
         /// <param name="position">The target position in the container</param>
         /// <returns>If the item was added</returns>
-        public bool AddItemPosition(Item item, Vector2Int position)
+        public bool AddItemPosition(ItemActor item, Vector2Int position)
```
that should stay item

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Containers/Container.cs:
```diff
@@ -225,7 +225,7 @@ public bool AddItemPosition(Item item, Vector2Int position)
         /// <param name="item">The item to add</param>
         /// <param name="position">Where the item should go, make sure this position is valid and free!</param>
-        private void AddItemUnchecked(Item item, Vector2Int position)
+        private void AddItemUnchecked(ItemActor item, Vector2Int position)
```
that should stay item

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Containers/Container.cs:
```diff
@@ -357,7 +357,7 @@ public bool IsAreaFree(RectInt area)
         /// <param name="item">The item to exclude from the check</param>
         /// <returns>If the given area is free</returns>
-        public bool IsAreaFreeExcluding(RectInt area, Item item)
+        public bool IsAreaFreeExcluding(RectInt area, ItemActor item)
```
that should stay item

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Containers/Container.cs:
```diff
@@ -382,7 +382,7 @@ public bool IsAreaFreeExcluding(RectInt area, Item item)
         /// </summary>
         /// <param name="item">The item to remove</param>
-        public void RemoveItem(Item item)
+        public void RemoveItem(ItemActor item)
```
that should stay item

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Containers/Container.cs:
```diff
@@ -430,7 +430,7 @@ private bool MoveItemUnchecked(StoredItem item)
         /// <param name="position">The position to check</param>
         /// <returns>The item at the position, or null if there is none</returns>
-        public Item ItemAt(Vector2Int position)
+        public ItemActor ItemAt(Vector2Int position)
```
that should stay item

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Containers/Container.cs:
```diff
@@ -449,7 +449,7 @@ public Item ItemAt(Vector2Int position)
         /// <param name="item">The item to look for</param>
         /// <returns>The item's position or (-1, -1)</returns>
-        public Vector2Int PositionOf(Item item)
+        public Vector2Int PositionOf(ItemActor item)
```
that should stay item

stilnat reviewed (commented):
pretty much just renaming item to itemActor since last review, good.

stilnat committed:
move synclist from Item to ItemActor, add constructors for item, remove dangerous empty constructor

stilnat committed:
remove unity logic from container and put it back in attachedContainer

stilnat committed:
remove itemActor from item

stilnat committed:
fix small issue with null red

stilnat committed:
fix some item stuff

stilnat committed:
remove synclist of traits for now

stilnat committed:
implement setcontainerunchecked

stilnat committed:
fix bug with item staying attached when left clicking

stilnat committed:
Greatly simplifies  setting container

The old way was very prone to break, I guess it was done for performance reasons, but you had to send a bool to say if it was already in the container or already removed. Here, I simply check that in SetContainer. Slower but sturdy. We'll optimise on that eventually. Also remove the dangerous "setting container unchecked" method. Surely efficient sometimes, but it's public and very prone to break too. That's end game optimisation.

stilnat committed:
Merge remote-tracking branch 'upstream/develop' into pr/1140

stilnat committed:
test equality by checking if actors are the same.

stilnat committed:
little fix

stilnat committed:
add describe method for item

stilnat committed:
remove script on human prefab to see all containers.

stilnat committed:
sync container using a rpc, correct method name

stilnat committed:
replace syncing using rpc with syncvar syncing

stilnat committed:
sync traits list in item

stilnat committed:
add item serializer

stilnat committed:
Merge remote-tracking branch 'upstream/develop' into pr/1140

stilnat committed:
make some field readOnly, clean up constructors in item

TaylorNAlbarnaz committed:
Added Item and Trait console commands

TaylorNAlbarnaz committed:
Added item.describe command and removed Update from actor

stilnat committed:
implement delete

stilnat committed:
Merge branch 'feature/humble-items' of https://github.com/TaylorNAlbarnaz/SS3D into pr/1140

stilnat committed:
fix container syncing using syncvar

stilnat committed:
make scripts called set container only if item.Actor exists

stilnat committed:
fix stupid thing in setContainer design

stilnat committed:
remove useless warning

stilnat committed:
some cleanup in itemActor

stilnat commented on Assets/Scripts/IngameDebugConsole/Scripts/Commands/ItemCommands.cs:
```diff
@@ -0,0 +1,89 @@
+using UnityEngine;
+using SS3D.Systems;
+
+namespace IngameDebugConsole.Commands
```
those command should probably be inheriting from Command, and be in the appropriate folder. The command class is integrated with our own systems, the IngameDebugConsole is an external asset. Sorry about the confusion ...

stilnat commented on Assets/Scripts/IngameDebugConsole/Scripts/DebugLogManager.cs:
```diff
@@ -416,8 +416,10 @@ private void Awake()
-				Destroy( gameObject );
-				return;
+#pragma warning disable COIMBRA0008 // Object.Destroy should not be used with Objects that can be an Actor.
+                Destroy( gameObject );
```
it's probably worth understanding why this warning is there.

stilnat commented on Assets/Scripts/SS3D/Systems/Interactions/InteractionController.cs:
```diff
@@ -114,10 +114,10 @@ private void HandleUse(InputAction.CallbackContext callbackContext)
             }
 
-            Item item = hands.ItemInHand;
+            ItemActor.Item item = hands.ItemInHand;
```
get actor directly here, only actor is used after

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Containers/Container.cs:
```diff
@@ -538,11 +531,6 @@ public bool ContainsItem(Item item)
         public bool CanStoreItem(Item item)
         {
-            // Do not store if the item is the container itself
-            if (AttachedTo.GetComponent<Item>() == item)
```
maybe put that back

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/Containers/Hands.cs:
```diff
@@ -80,19 +83,19 @@ protected override void OnDestroyed()
             }
 
-            if (item.Container != SelectedHandContainer && item.Container != null)
+            if (item.GetItem.Container != SelectedHandContainer && item.GetItem.Container != null)
```
add a Container property on itemActor to avoid that ?

stilnat commented on Assets/Scripts/SS3D/Systems/Traits/Trait.cs:
```diff
@@ -9,7 +9,7 @@ namespace SS3D.Systems
     {
         // Hash used for identification
-        protected int hash;
+        public int hash;
```
that was quickly changed for syncing but that's a mistake, instead, use a serializer for traits, getting those field with public getter properties.

stilnat commented on Assets/Scripts/SS3D/Systems/Inventory/UI/SingleItemContainerSlot.cs:
```diff
@@ -65,7 +66,10 @@ public override void OnItemDisplayDrop(ItemDisplay display)
         private void UpdateDisplay()
         {
-            ItemDisplay.Item = _container.Container.Items.FirstOrDefault();
+            var item = _container.Container.Items.FirstOrDefault();
```
that surely could be shortened

TaylorNAlbarnaz committed:
Merge branch 'main' into feature/humble-items

TaylorNAlbarnaz committed:
Tests for Containers and Items

stilnat committed:
add incomplete trait serializer

TaylorNAlbarnaz committed:
Merge branch 'main' into feature/humble-items

TaylorNAlbarnaz committed:
Fixes NukeDisk conflict

stilnat committed:
add trait serializer

stilnat committed:
Merge branch 'feature/humble-items' of https://github.com/TaylorNAlbarnaz/SS3D into pr/1140

stilnat committed:
remove itemactor,  item only

stilnat committed:
update some item with new item class

stilnat committed:
mod fishnet

stilnat committed:
make proper init method, and move setContainer logic

stilnat committed:
rename IsOnContainer to InContainer

stilnat committed:
rename initialised to _initialised

stilnat committed:
add check for trait adding

stilnat committed:
put private member protected in trait like before

stilnat committed:
add some attributes and docs

stilnat committed:
adds more attributes

stilnat committed:
add a few more attributes

stilnat committed:
fix validate size, some cleanup

stilnat committed:
fix init method

stilnat committed:
update item commands

stilnat committed:
add basic doc
